app/src/main/java/emu/skyline/loader/RomFile.kt

```diff
@@ -123,16 +124,81 @@ class AppEntry : Serializable {
 }
 
 /**
- * This class is used as the base class for all loaders
+ * This class is used as interface between libskyline and Kotlin for loaders
  */
-internal abstract class BaseLoader(val context : Context, val format : RomFormat) {
+internal class RomFile(val context : Context, val format : RomFormat, val file : ParcelFileDescriptor) : AutoCloseable {
     /**
-     * This is used to get the [AppEntry] for the specified [file] at the supplied [uri]
+     * This is a pointer to the corresponding C++ Loader class
      */
-    abstract fun getAppEntry(file : RandomAccessDocument, uri : Uri) : AppEntry
+    var instance : Long
+
+    init {
+        System.loadLibrary("skyline")
+
+        instance = initialize(format.ordinal, file.fd)
+    }
 
     /**
-     * This returns if the supplied [file] is a valid ROM or not
+     * This allocates and initializes a new loader object
+     * @param format The format of the ROM
+     * @param romFd A file descriptor of the ROM
+     * @return A pointer to the newly allocated object, or 0 if the ROM is invalid
      */
-    abstract fun verifyFile(file : RandomAccessDocument) : Boolean
-}
+    private external fun initialize(format : Int, romFd : Int) : Long
+
+    /**
+     * @return Whether the ROM contains assets, such as an icon or author information
+     */
+    private external fun hasAssets(instance : Long) : Boolean
+
+    /**
+     * @return A ByteArray containing the application's icon as a bitmap
+     */
+    private external fun getIcon(instance : Long) : ByteArray
+
+    /**
+     * @return A String containing the name of the application
+     */
+    private external fun getApplicationName(instance : Long) : String
+
+    /**
+     * @return A String containing the publisher of the application
+     */
+    private external fun getApplicationPublisher(instance : Long) : String
+
+    /**
+     * This destroys an existing loader object and frees it's resources
+     */
+    private external fun destroy(instance : Long)
+
+    /**
+     * This is used to get the [AppEntry] for the specified NRO
+     */
+    fun getAppEntry(uri : Uri) : AppEntry {
+        return if (hasAssets(instance)) {
+            val rawIcon = getIcon(instance)
+            val icon = BitmapFactory.decodeByteArray(rawIcon, 0, rawIcon.size)
```

Comment:
You should only do this if the size isn't 0, otherwise set `icon` to `null`.
```kotlin
val icon = if (rawIcon.size != 0) BitmapFactory.decodeByteArray(rawIcon, 0, rawIcon.size) else null
```

Reply:
That's effectively what HasAssets does, all programs with nacp data also have an icon afaik

Reply:
They can have an icon `ASET` header but the icon itself can be empty and the size can be `0`. So, you should add in the check.